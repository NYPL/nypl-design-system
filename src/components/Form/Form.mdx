import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as FormStories from "./Form.stories";

<Meta of={FormStories} />

# Form

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.23.2`   |
| Latest            | `1.0.6`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Spacing Variants](#spacing-variants)
- [Example Code](#example-code)

## Overview

<Description of={FormStories} />

The `Form` component renders a standard `<form>` element and should be used to
handle layout and spacing for child input fields. `FormRow` and `FormField`
components should be used to build the `<form>` structure and to arrange input
fields as needed.

<Source
  code={`
<Form id="form-id">
  <FormRow>
    <FormField>{/* ... */}</FormField>
  </FormRow>
  <FormRow>
    <FormField>{/* ... */}</FormField>
  </FormRow>
</Form>
`}
  language="tsx"
/>

`FormField` should be used as a parent for all input components from the
Reservoir Design System (DS) (`Button`, `Select`, `TextInput`, etc.).

`FormRow` should be used as a parent of multiple `FormField` components when you
need to render multiple input components in a horizontal row.

## Component Props

<Canvas of={FormStories.WithControls} />

<ArgTypes of={FormStories.WithControls} />

## Accessibility

The `Form` component renders a standard `<form>` element and should be used to
wrap all form elements. A `Form` should have a `Button` component with a `type`
of `submit` to submit the form to a server through the `Form`'s `action` prop.

Resources:

- [WebAIM Creating Accessible Forms](https://webaim.org/techniques/forms/)
- [Deque The Anatomy of Accessible Forms: Best Practices](https://www.deque.com/blog/anatomy-of-accessible-forms-best-practices/)

## Spacing Variants

By default, the `Form` component will handle the NYPL spacing around form input
elements. The default spacing value is large `grid.l`, which corresponds to the
CSS variable `--nypl-space-l` (2rem / 32px).

**IMPORTANT:** The default spacing should not be overwritten without a very good
reason.

Below are examples of the spacing values available to use in the `gap` prop of
the `Form` component: `"grid.xxs"`, `"grid.xs"`, `"grid.s"`, `"grid.m"`,
`"grid.l"`, `"grid.xl"`, `"grid.xxl"`.

<Canvas of={FormStories.FormSpacing} />

## Example Code

<Source
  code={`
<Form action="/end/point" method="get" gap="grid.l" id="example-form">
  <FormField>
    <TextInput
      labelText="Username"
      helperText="Make sure to complete this field."
      isRequired
    />
  </FormField>
  <FormField>
    <TextInput
      labelText="Password"
      helperText="Make sure to complete this field."
      isRequired
    />
  </FormField>
  <FormRow>
    <FormField>
      <TextInput
        labelText="Phone Field"
        helperText="This one is up to you."
        type="tel"
      />
    </FormField>
    <FormField>
      <TextInput
        labelText="URL Field"
        helperText="This one is up to you."
        type="url"
      />
    </FormField>
    <FormField>
      <TextInput
        labelText="Age"
        helperText="This one is up to you."
        type="number"
      />
    </FormField>
  </FormRow>
  <FormField>
    <Button>Submit</Button>
  </FormField>
</Form>
`}
  language="tsx"
/>
