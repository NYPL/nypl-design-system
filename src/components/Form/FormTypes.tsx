import { GridGaps as FormSpacing } from "../Grid/GridTypes";

export { FormSpacing };
