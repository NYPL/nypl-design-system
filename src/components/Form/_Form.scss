.form {
  @include space-stack-s;
}

.form-field {
  @include space-stack-l;
}

.form-row {
  @include space-stack-l;

  display: flex;

  .form-field {
    flex-grow: 1;
    margin-bottom: 0;

    &:not(:last-child) {
      @include space-inline-l;
    }
  }
}
