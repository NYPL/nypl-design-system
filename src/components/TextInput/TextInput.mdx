import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as TextInputStories from "./TextInput.stories";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";
import Link from "../Link/Link";
import { changelogData } from "./textInputChangelogData";

<Meta of={TextInputStories} />

# TextInput

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.22.0`   |
| Latest            | `2.1.0`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#labelling-variations" target="_self">Labelling Variations</Link>}
- {<Link href="#browser-states" target="_self">Browser States</Link>}
- {<Link href="#isclearable-button" target="_self">isClearable Button</Link>}
- {<Link href="#number-type" target="_self">Number Type</Link>}
- {<Link href="#html-in-helper-text" target="_self">HTML in Helper Text</Link>}
- {<Link href="#textarea" target="_self">Textarea</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={TextInputStories} />

This component renders a text input form element. See below for configuration
information. For optimal accessibility, the `labelText` property is a required
prop, regardless of the label visibility. Additionally, while the `id` prop is
optional, a unique `id` attribute is necessary for accessibility. If the prop
is left blank, a value will be generated for you.

## Component Props

<Canvas of={TextInputStories.WithControls} />

<ArgTypes of={TextInputStories.WithControls} />

## Accessibility

The `TextInput` component renders an HTML `<input>` element which is accessible
via keyboard and screen reader. The `aria-required` and `required` attributes are
set with the `isRequired` prop. The `aria-disabled` and `disabled` attributes are
set with the `isDisabled` prop. The `aria-invalid` is set with the `isInvalid`
prop.

Internally, a `Label` is associated with the `<input>` element. When `showLabel`
is set to false, the `<input>` element's `aria-label` attribute is set to the
required `labelText` value.

The `helperText` and the `invalidText` are associated with the `<input>` element
through the `aria-describedby` attribute.

When the `isClearable` prop is set to `true`, as text is added to the input
field a close button is rendered to clear the input value. Once the close button
is clicked, the input value will be cleared, the close button will become hidden,
and focus will be set to the input element.

When the `type` prop is set to `"textarea"`, the `<textarea>` element
is rendered instead of the `<input>` element. This element follows all the same
accessibility rules described above.

If the `type` prop is set to `"email"`, `"tel"`, or `"url"`, the appropriate
`autocomplete` attribute will be added. The autocomplete attribute makes inputs
easier and more efficient to complete for all users.

Resources:

- [MDN input: The Input (Form Input) element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input)
- [MDN textarea: The Textarea element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea)
- [MDN HTML attribute: Autocomplete](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete)
- [Chakra UI Input](https://chakra-ui.com/docs/components/form/input)
- [Chakra UI Textarea](https://chakra-ui.com/docs/components/form/textarea)

## Labelling Variations

A TextInput can be rendered with or without visible labels. When `showLabel` is
set to `false`, an `aria-label` attribute is added to the input/textarea element
to maintain accessibility. If the component needs to be required, the
`showRequiredLabel` prop can be used to show or hide the "(Required)" text
within the `label` element.

<Canvas of={TextInputStories.LabellingVariations} />

## Browser States

<Canvas of={TextInputStories.BrowserStates} />

## isClearable Button

Set the `isClearable` prop to `true` to render a button that clears the input
value. This close button will render when the input has a value and disappear
when the input is empty, including after it is clicked.

This button has visually hidden text, a "close" icon, and renders on top
of the input field on the right side. This clear button does not render when
the `type` prop is set to `"textarea"`.

Note: This works best when the `TextInput` is used as an uncontrolled component.

<Canvas of={TextInputStories.isClearableButton} />

### Controlled isClearableCallback Example

A `TextInput` component is controlled when the `value` and `onChange` props are
set. For a controlled component, the `isClearableCallback` prop can be used to
clear the input value in the consuming application. This callback is only called
when the close button is clicked.

<Source
  code={`
// ExampleControlled.js
export const App = () => {
  const [value, setValue] = React.useState("");
  return (
    <TextInput
      id="isClearable-controlled-example"
      isClearable
      isClearableCallback={() => setValue("")}
      labelText="What is your favorite color?"
      onChange={(e) => setValue(e.target.value)}
      placeholder="i.e. blue, green, etc."
      value={value}
    />
  );
};
`}
  language="jsx"
/>

<Canvas of={TextInputStories.ControlledExample} />

## Number Type

The `TextInput` component can be configured to render a number input by setting
`type="number"`. In this input type, the `maxLength` prop does not restrict the
length of the number that is passed to the input. This is the expected behavior
as the native HTML input and `maxlength` attribute behave this way.

For this type, the `min` and `max` props do restrict the value of the number
input that is passed. This can be seen by using the up and down arrows to
increase or decrease the number value, respectively. However, if the `max` prop
is set to "200", for example, the user can still enter a number such as "400".
The consuming application must handle validating the entered number value.

If the `min` prop value is greater than the `max` prop value, it will trigger
the invalid state.

<Canvas of={TextInputStories.NumberType} />

### Alternate Number Pattern

Another pattern for entering numbers in a restrictive way is to use the
`maxLength` and `pattern` props. Together, these props can be set to define how
many digits and what characters are allowed. For example, if a zip code is
required, the `maxLength` prop can be set to "5" and the `pattern` prop can be
set to `"[0-9]\*"` to allow only numbers to be entered and limit the number of
characters to 5.

Note that the `type` prop is not set to "number" in this example and that the
consuming application must handle validating the value entered. Browsers will
also provide their own validation for the input.

<Source
  code={`
<TextInput
  helperText="Enter a 5-digit zip code."
  id="number-pattern"
  labelText="Enter a zip code"
  maxLength={5}
  pattern="[0-9]*"
  placeholder="i.e. 10018"
/>
`}
  language="jsx"
/>

<Canvas of={TextInputStories.AlternateNumberPattern} />

## HTML in Helper Text

HTML can be passed into the `helperText` prop as a string or HTML.

<Source
  code={`
helperText="Choose <b>wisely!</b>"
// or
helperText={<>Choose <b>wisely!</b></>}
`}
  language="jsx"
/>

<Canvas of={TextInputStories.HTMLHelperText} />

## Textarea

The TextInput component includes a multiline `textarea` form field. To render a
`textarea` element, pass `"textarea"` for the `type` prop. All the variations
described above are available for the `textarea` option.

<Canvas of={TextInputStories.Textarea} />

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
