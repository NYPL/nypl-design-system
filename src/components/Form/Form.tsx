import * as React from "react";

import { FormGaps } from "./FormTypes";
import SimpleGrid from "../Grid/SimpleGrid";
import generateUUID from "../../helpers/generateUUID";

interface FormBaseProps {
  /** className to be applied to FormRow, FormField, and Form */
  className?: string;
  /** Optional spacing size; if omitted, the default `large` (2rem / 32px)
   * spacing will be used; ```IMPORTANT: for general form layout, this prop
   * should not be used``` */
  gap?: FormGaps;
  /** ID that other components can cross reference (internal use) */
  id?: string;
}

export interface FormChildProps extends FormBaseProps {}

export interface FormProps extends FormBaseProps {
  /** Optional form `action` attribute */
  action?: string;
  /** Optional className you can add in addition to `form` */
  className?: string;
  /** Optional ID that other components can cross reference */
  id?: string;
  /** Sets the `enctype` attribute to "multipart/form-data" for file uploads. */
  isFileUploader?: boolean;
  /** Optional form `method` attribute */
  method?: "get" | "post";
  /** Function to call for the `onSubmit` form event. */
  onSubmit?: (e: React.FormEvent<HTMLFormElement>) => void;
}

/** FormRow child-component */
export function FormRow(props: React.PropsWithChildren<FormChildProps>) {
  const { children, className, gap, id } = props;
  const count = React.Children.count(children);
  const alteredChildren = React.Children.map(
    children,
    (child: React.ReactElement, i) => {
      if (!child) return null;
      if (child.type === FormField || child.props.mdxType === "FormField") {
        return React.cloneElement(child, { id: `${id}-grandchild${i}` });
      }
      console.warn("FormRow children must be `FormField` components.");
      return null;
    }
  );
  return (
    <SimpleGrid columns={count} className={className} gap={gap} id={id}>
      {alteredChildren}
    </SimpleGrid>
  );
}

/** FormField child-component */
export function FormField(props: React.PropsWithChildren<FormChildProps>) {
  const { children, className, gap, id } = props;
  return (
    <SimpleGrid columns={1} className={className} gap={gap} id={id}>
      {children}
    </SimpleGrid>
  );
}

/** Main Form component */
export default function Form(props: React.PropsWithChildren<FormProps>) {
  const {
    action,
    children,
    className,
    gap = FormGaps.Large,
    id = generateUUID(),
    isFileUploader = false,
    method,
    onSubmit,
  } = props;

  let attributes = {};
  action && (attributes["action"] = action);

  method &&
    (method === "get" || method === "post") &&
    (attributes["method"] = method);

  isFileUploader && (attributes["enctype"] = "multipart/form-data");

  const alteredChildren = React.Children.map(
    children,
    (child: React.ReactElement, i) => {
      return React.cloneElement(child, { gap, id: `${id}-child${i}` });
    }
  );

  return (
    <form
      aria-label="form"
      className={className}
      id={id}
      onSubmit={onSubmit}
      {...attributes}
    >
      <SimpleGrid columns={1} gap={gap} id={`${id}-parent`}>
        {alteredChildren}
      </SimpleGrid>
    </form>
  );
}
