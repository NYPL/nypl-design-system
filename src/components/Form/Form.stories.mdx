import { Flex, Spacer } from "@chakra-ui/react";
import { action } from "@storybook/addon-actions";
import {
  ArgsTable,
  Canvas,
  Description,
  Meta,
  Story,
} from "@storybook/addon-docs";
import { withDesign } from "storybook-addon-designs";

import Button from "../Button/Button";
import Checkbox from "../Checkbox/Checkbox";
import CheckboxGroup from "../CheckboxGroup/CheckboxGroup";
import DatePicker from "../DatePicker/DatePicker";
import { DatePickerTypes } from "../DatePicker/DatePickerTypes";
import Form, { FormRow, FormField } from "./Form";
import { FormGaps } from "./FormTypes";
import Heading from "../Heading/Heading";
import { HeadingLevels } from "../Heading/HeadingTypes";
import HorizontalRule from "../HorizontalRule/HorizontalRule";
import Label from "../Label/Label";
import Radio from "../Radio/Radio";
import RadioGroup from "../RadioGroup/RadioGroup";
import Select from "../Select/Select";
import TextInput from "../TextInput/TextInput";
import { getCategory } from "../../utils/componentCategories";
import SimpleGrid from "../Grid/SimpleGrid";
import { getStorybookEnumValues } from "../../utils/utils";

export const enumValues = getStorybookEnumValues(FormGaps, "FormGaps");

<Meta
  title={getCategory("Form")}
  component={Form}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=11895%3A547",
    },
    jest: ["Form.test.tsx"],
  }}
  argTypes={{
    action: { control: false },
    id: { control: false },
    method: {
      control: { type: "radio" },
      options: ["get", "post"],
    },
    gap: {
      control: { type: "select" },
      table: { defaultValue: { summary: "FormGaps.Large" } },
      options: enumValues.options,
    },
  }}
/>

# Form

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.23.2`   |
| Latest            | `0.26.0`   |

<Description of={Form} />

The `Form` component renders a standard `<form>` element and should be used to
handle layout and spacing for child input fields. `FormRow` and `FormField`
components should be used to build the `<form>` structure and to arrange input
fields as needed.

```jsx
<Form>
  <FormRow>
    <FormField>{/* ... */}</FormField>
  </FormRow>
  <FormRow>
    <FormField>{/* ... */}</FormField>
  </FormRow>
</Form>
```

`FormField` should be used as a parent for all input components from the DS
(`Button`, `Select`, `TextInput`, etc.).

`FormRow` should be used as a parent of multiple `FormField` components when you
need to render multiple input components in a horizontal row.

<Canvas withToolbar>
  <Story
    name="Form with Controls"
    args={{
      action: "/end/point",
      className: undefined,
      id: "form-id",
      method: "get",
      gap: "FormGaps.Large",
    }}
  >
    {(args) => (
      <Form {...args} gap={enumValues.getValue(args.gap)}>
        <FormRow>
          <FormField>
            <TextInput
              labelText="First Name"
              helperText="Make sure to complete this field."
              isRequired
            />
          </FormField>
          <FormField>
            <TextInput
              labelText="Last Name"
              helperText="Make sure to complete this field."
              isRequired
            />
          </FormField>
          <FormField>
            <DatePicker
              id="date-range"
              dateType={DatePickerTypes.Full}
              dateFormat="yyyy-MM-dd"
              helperTextFrom="From this date."
              helperTextTo="To this date."
              helperText="Select a valid date range."
              invalidText="Please select a valid date range."
              isDateRange
              labelText="Select the date range you want to visit NYPL"
              minDate="1/1/2021"
              maxDate="1/1/2022"
              nameFrom="visit-dates"
              showLabel={false}
            />
          </FormField>
        </FormRow>
        <FormField>
          <TextInput
            labelText="Username"
            helperText="Make sure to complete this field."
            isRequired
          />
        </FormField>
        <FormField>
          <TextInput
            labelText="Password"
            helperText="Make sure to complete this field."
            isRequired
          />
        </FormField>
        <FormRow>
          <FormField>
            <TextInput
              labelText="Phone Field"
              helperText="This one is up to you."
              type="tel"
            />
          </FormField>
          <FormField>
            <TextInput
              labelText="URL Field"
              helperText="This one is up to you."
              type="url"
            />
          </FormField>
          <FormField>
            <TextInput
              labelText="Age"
              helperText="This one is up to you."
              type="number"
            />
          </FormField>
        </FormRow>
        <FormRow>
          <FormField>
            <CheckboxGroup
              isFullWidth
              labelText="Checkbox Group"
              name="checkbox-example"
              optReqFlag={false}
            >
              <Checkbox
                labelText={
                  <Flex>
                    <span>Arts</span>
                    <Spacer />
                    <span>4</span>
                  </Flex>
                }
                value="arts"
              />
              <Checkbox
                labelText={
                  <Flex>
                    <span>English</span>
                    <Spacer />
                    <span>23</span>
                  </Flex>
                }
                value="English"
              />
              <Checkbox
                labelText={
                  <Flex>
                    <span>Science</span>
                    <Spacer />
                    <span>10</span>
                  </Flex>
                }
                value="Science"
              />
              <Checkbox
                labelText={
                  <Flex>
                    <span>Math</span>
                    <Spacer />
                    <span>3</span>
                  </Flex>
                }
                value="Math"
              />
            </CheckboxGroup>
          </FormField>
          <FormField>
            <RadioGroup labelText="Radio Group" name="rg1" optReqFlag={false}>
              <Radio labelText="Radio 1" showLabel />
              <Radio labelText="Radio 2" showLabel />
              <Radio labelText="Radio 3" showLabel />
              <Radio labelText="Radio 4" showLabel />
            </RadioGroup>
          </FormField>
          <FormField>
            <Select
              helperText="The select field helper text."
              labelText="Select Field"
              showLabel={true}
            >
              <option>Option 1</option>
              <option>Option 2</option>
              <option>Option 3</option>
              <option>Option 4</option>
              <option>Option 5</option>
            </Select>
          </FormField>
        </FormRow>
        <FormField>
          <Button>Submit</Button>
        </FormField>
      </Form>
    )}
  </Story>
</Canvas>

<ArgsTable story="Form with Controls" />

## Spacing Variants

export const formRow = (nameString, size) => {
  const labelText = `Size: ${nameString} (${size})`;
  return (
    <li key={size}>
      <Heading level={HeadingLevels.Three}>{labelText}</Heading>
      <Form gap={size}>
        <FormRow>
          <FormField>
            <Select
              helperText="The select field helper text."
              labelText="Select Field"
              showLabel={true}
            >
              <option>Option 1</option>
              <option>Option 2</option>
              <option>Option 3</option>
              <option>Option 4</option>
            </Select>
          </FormField>
          <FormField>
            <Select
              helperText="The select field helper text."
              labelText="Select Field"
              showLabel={true}
            >
              <option>Option 1</option>
              <option>Option 2</option>
              <option>Option 3</option>
              <option>Option 4</option>
            </Select>
          </FormField>
          <FormField>
            <Select
              helperText="The select field helper text."
              labelText="Select Field"
              showLabel={true}
            >
              <option>Option 1</option>
              <option>Option 2</option>
              <option>Option 3</option>
              <option>Option 4</option>
            </Select>
          </FormField>
        </FormRow>
      </Form>
      <HorizontalRule />
    </li>
  );
};
export const sizes = [];
for (const formGap in FormGaps) {
  sizes.push(formRow(`FormGaps.${formGap}`, FormGaps[formGap]));
}
export const getForms = (list) => <ul style={{ listStyle: "none" }}>{list}</ul>;

By default, the `Form` component will handle the NYPL spacing around form input
elements. The default spacing value is `large`, which corresponds to the CSS
variable `--nypl-space-l` (2rem / 32px).

**IMPORTANT:** The default spacing should not be overwritten without a very good reason.

Below are the spacing variants available with the `FormGaps` enum.

<Canvas>
  <Story
    name="Spacing Variants"
    args={{
      id: "spacing-form-id",
    }}
  >
    {getForms(sizes)}
  </Story>
</Canvas>

## Example Code

<Story name="Example Code" />

```jsx
<Form action="/end/point" method="get" gap={FormGaps.Large}>
  <FormField>
    <TextInput
      labelText="Username"
      helperText="Make sure to complete this field."
      isRequired
    />
  </FormField>
  <FormField>
    <TextInput
      labelText="Password"
      helperText="Make sure to complete this field."
      isRequired
    />
  </FormField>
  <FormRow>
    <FormField>
      <TextInput
        labelText="Phone Field"
        helperText="This one is up to you."
        type="tel"
      />
    </FormField>
    <FormField>
      <TextInput
        labelText="URL Field"
        helperText="This one is up to you."
        type="url"
      />
    </FormField>
    <FormField>
      <TextInput
        labelText="Age"
        helperText="This one is up to you."
        type="number"
      />
    </FormField>
  </FormRow>
  <FormField>
    <Button>Submit</Button>
  </FormField>
</Form>
```
