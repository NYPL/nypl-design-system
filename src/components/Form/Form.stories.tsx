import { Flex, Spacer } from "@chakra-ui/react";
import type { Meta, StoryObj } from "@storybook/react";

import Button from "../Button/Button";
import ButtonGroup from "../ButtonGroup/ButtonGroup";
import Checkbox from "../Checkbox/Checkbox";
import CheckboxGroup from "../CheckboxGroup/CheckboxGroup";
import DatePicker from "../DatePicker/DatePicker";
import Form, { FormRow, FormField } from "./Form";
import { gridGapsArray } from "../Grid/SimpleGrid";
import Heading from "../Heading/Heading";
import HorizontalRule from "../HorizontalRule/HorizontalRule";
import Radio from "../Radio/Radio";
import RadioGroup from "../RadioGroup/RadioGroup";
import Select from "../Select/Select";
import TextInput from "../TextInput/TextInput";

const meta: Meta<typeof Form> = {
  title: "Components/Form Elements/Form",
  component: Form,
  argTypes: {
    action: { control: false },
    id: { control: false },
    method: {
      control: { type: "radio" },
      options: ["get", "post"],
    },
    gap: {
      control: { type: "select" },
      table: { defaultValue: { summary: "grid.l" } },
      options: gridGapsArray,
    },
  },
};

export default meta;
type Story = StoryObj<typeof Form>;

/**
 * Main Story for the Form component. This must contains the `args`
 * and `parameters` properties in this object.
 */
export const WithControls: Story = {
  args: {
    action: "/end/point",
    className: undefined,
    id: "form-id",
    method: "get",
    gap: "grid.l",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?type=design&node-id=10734-2768",
    },
    jest: "Form.test.tsx",
  },
  render: (args) => (
    <Form {...args} id="form-id">
      <FormRow>
        <FormField>
          <TextInput
            helperText="Make sure to complete this field."
            id="first-name"
            isRequired
            labelText="First Name"
          />
        </FormField>
        <FormField>
          <TextInput
            helperText="Make sure to complete this field."
            id="last-name"
            isRequired
            labelText="Last Name"
          />
        </FormField>
        <FormField>
          <DatePicker
            dateType="full"
            dateFormat="yyyy-MM-dd"
            helperTextFrom="From this date."
            helperTextTo="To this date."
            helperText="Select a valid date range."
            id="date-range"
            invalidText="Please select a valid date range."
            isDateRange
            labelText="Select the date range you want to visit NYPL"
            minDate="1/1/2021"
            maxDate="1/1/2022"
            nameFrom="visit-dates"
            showLabel={false}
          />
        </FormField>
      </FormRow>
      <FormField>
        <TextInput
          helperText="Make sure to complete this field."
          id="username"
          isRequired
          labelText="Username"
        />
      </FormField>
      <FormField>
        <TextInput
          helperText="Make sure to complete this field."
          id="password"
          isRequired
          labelText="Password"
        />
      </FormField>
      <FormRow>
        <FormField>
          <TextInput
            helperText="This one is up to you."
            id="phone"
            labelText="Phone Field"
            type="tel"
          />
        </FormField>
        <FormField>
          <TextInput
            helperText="This one is up to you."
            id="url"
            labelText="URL Field"
            type="url"
          />
        </FormField>
        <FormField>
          <TextInput
            helperText="This one is up to you."
            id="age"
            labelText="Age"
            type="number"
          />
        </FormField>
      </FormRow>
      <FormRow>
        <FormField>
          <CheckboxGroup
            id="checkbox-group"
            isFullWidth
            labelText="Checkbox Group"
            name="checkbox-example"
          >
            <Checkbox
              id="arts"
              labelText={
                <Flex>
                  <span>Arts</span>
                  <Spacer />
                  <span>4</span>
                </Flex>
              }
              value="arts"
            />
            <Checkbox
              id="english"
              labelText={
                <Flex>
                  <span>English</span>
                  <Spacer />
                  <span>23</span>
                </Flex>
              }
              value="English"
            />
            <Checkbox
              id="science"
              labelText={
                <Flex>
                  <span>Science</span>
                  <Spacer />
                  <span>10</span>
                </Flex>
              }
              value="Science"
            />
            <Checkbox
              id="math"
              labelText={
                <Flex>
                  <span>Math</span>
                  <Spacer />
                  <span>3</span>
                </Flex>
              }
              value="Math"
            />
          </CheckboxGroup>
        </FormField>
        <FormField>
          <RadioGroup id="radio-group" labelText="Radio Group" name="rg1">
            <Radio id="radio1" labelText="Radio 1" value="radio1" />
            <Radio id="radio2" labelText="Radio 2" value="radio2" />
            <Radio id="radio3" labelText="Radio 3" value="radio3" />
            <Radio id="radio4" labelText="Radio 4" value="radio4" />
          </RadioGroup>
        </FormField>
        <FormField>
          <Select
            helperText="The select field helper text."
            id="select"
            labelText="Select Field"
            name="select"
            showLabel={true}
          >
            <option>Option 1</option>
            <option>Option 2</option>
            <option>Option 3</option>
            <option>Option 4</option>
            <option>Option 5</option>
          </Select>
        </FormField>
      </FormRow>
      <FormRow>
        <FormField>
          <ButtonGroup>
            <Button id="submit">Submit</Button>
          </ButtonGroup>
        </FormField>
      </FormRow>
    </Form>
  ),
};

// The following functions are used in the remaining Form example stories
const formRow = (gapValue) => {
  const labelText = `Size: ${gapValue}`;
  return (
    <li key={gapValue}>
      <Heading level="h3" size="heading6">
        {labelText}
      </Heading>
      <Form gap={gapValue} id={`form-spacing-${gapValue}`}>
        <FormRow>
          <FormField>
            <Select
              helperText="The select field helper text."
              id={`select-spacing-${gapValue}-1`}
              labelText="Select Field"
              name={`select-spacing-${gapValue}-1`}
              showLabel={true}
            >
              <option>Option 1</option>
              <option>Option 2</option>
              <option>Option 3</option>
              <option>Option 4</option>
            </Select>
          </FormField>
          <FormField>
            <Select
              helperText="The select field helper text."
              id={`select-spacing-${gapValue}-2`}
              labelText="Select Field"
              name={`select-spacing-${gapValue}-2`}
              showLabel={true}
            >
              <option>Option 1</option>
              <option>Option 2</option>
              <option>Option 3</option>
              <option>Option 4</option>
            </Select>
          </FormField>
          <FormField>
            <Select
              helperText="The select field helper text."
              id={`select-spacing-${gapValue}-3`}
              labelText="Select Field"
              name={`select-spacing-${gapValue}-3`}
              showLabel={true}
            >
              <option>Option 1</option>
              <option>Option 2</option>
              <option>Option 3</option>
              <option>Option 4</option>
            </Select>
          </FormField>
        </FormRow>
      </Form>
      <HorizontalRule />
    </li>
  );
};
const sizes = [];
const gapValues = gridGapsArray;
for (const gap in gapValues) {
  sizes.push(formRow(gapValues[gap]));
}
const getForms = (list) => <ul style={{ listStyle: "none" }}>{list}</ul>;

// The following are additional Form example Stories.
export const FormSpacing: Story = {
  render: () => getForms(sizes),
};
