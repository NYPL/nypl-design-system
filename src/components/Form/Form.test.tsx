import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";

import Form, { FormRow, FormField } from "./Form";
import TextInput from "../TextInput/TextInput";

describe("Form", () => {
  //   let wrapper: Enzyme.ShallowWrapper<any, any>;
  it("Renders a <form> element", () => {
    const wrapper = Enzyme.shallow(<Form></Form>);
    expect(wrapper.find("form")).to.have.lengthOf(1);
  });

  it("Renders a <form> element with child .form-row element", () => {
    const wrapper = Enzyme.mount(
      <Form>
        <FormRow></FormRow>
      </Form>
    );
    expect(wrapper.find(".form")).to.have.lengthOf(1);
    expect(wrapper.find(".form .form-row")).to.have.lengthOf(1);
  });

  it("Renders a <form> element with child .form-field element", () => {
    const wrapper = Enzyme.mount(
      <Form>
        <FormField></FormField>
      </Form>
    );
    expect(wrapper.find(".form")).to.have.lengthOf(1);
    expect(wrapper.find(".form .form-field")).to.have.lengthOf(1);
  });

  it("Renders a <form> element with .form-row, .form-field and input elements properly nested", () => {
    const wrapper = Enzyme.mount(
      <Form>
        <FormRow>
          <FormField>
            <TextInput labelText="Input Field" />
          </FormField>
        </FormRow>
      </Form>
    );
    expect(wrapper.find(".form")).to.have.lengthOf(1);
    expect(wrapper.find(".form .form-row")).to.have.lengthOf(1);
    expect(wrapper.find(".form .form-row .form-field")).to.have.lengthOf(1);
    expect(
      wrapper.find(".form .form-row .form-field .textinput")
    ).to.have.lengthOf(1);
  });

  it("Renders a <form> element with .form-row and .form-field elements properly nested", () => {
    const wrapper = Enzyme.shallow(
      <Form action="/end/point" method="get"></Form>
    );
    expect(wrapper.find("form")).to.have.lengthOf(1);
    expect(wrapper.find("form").prop("action")).to.equal("/end/point");
    expect(wrapper.find("form").prop("method")).to.equal("get");
  });
});
