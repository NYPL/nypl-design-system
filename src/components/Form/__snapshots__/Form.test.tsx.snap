// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Snapshot Renders the UI snapshot correctly 1`] = `
<form
  aria-label="form"
  className="css-0"
  id="snapshot-form"
>
  <div
    className="css-fadkce"
    id="snapshot-form-parent"
  >
    <div
      className="css-dofnwd"
      id="snapshot-form-child0"
    >
      <div
        className="css-7ajbp2"
        id="snapshot-form-child0-grandchild0"
      >
        Form Field 1
      </div>
      <div
        className="css-7ajbp2"
        id="snapshot-form-child0-grandchild1"
      >
        Form Field 2
      </div>
      <div
        className="css-7ajbp2"
        id="snapshot-form-child0-grandchild2"
      >
        Form Field 3
      </div>
    </div>
  </div>
</form>
`;

exports[`Form Snapshot Renders the UI snapshot correctly 2`] = `
<form
  aria-label="form"
  className="css-qvgr6z"
  color="ui.error.primary"
  id="chakra"
  p="20px"
>
  <div
    className="css-fadkce"
    id="chakra-parent"
  >
    <div
      className="css-dofnwd"
      id="chakra-child0"
    >
      <div
        className="css-7ajbp2"
        id="chakra-child0-grandchild0"
      >
        Form Field 1
      </div>
      <div
        className="css-7ajbp2"
        id="chakra-child0-grandchild1"
      >
        Form Field 2
      </div>
      <div
        className="css-7ajbp2"
        id="chakra-child0-grandchild2"
      >
        Form Field 3
      </div>
    </div>
  </div>
</form>
`;

exports[`Form Snapshot Renders the UI snapshot correctly 3`] = `
<form
  aria-label="form"
  className="css-0"
  data-testid="props"
  id="props"
>
  <div
    className="css-fadkce"
    id="props-parent"
  >
    <div
      className="css-dofnwd"
      id="props-child0"
    >
      <div
        className="css-7ajbp2"
        id="props-child0-grandchild0"
      >
        Form Field 1
      </div>
      <div
        className="css-7ajbp2"
        id="props-child0-grandchild1"
      >
        Form Field 2
      </div>
      <div
        className="css-7ajbp2"
        id="props-child0-grandchild2"
      >
        Form Field 3
      </div>
    </div>
  </div>
</form>
`;
